system_metrics/basic_sender_test.go

```diff
@@ -0,0 +1,370 @@
+	}, nil
+}
+
+type mockServer struct {
+	createErr bool
+	addErr    bool
+	closeErr  bool
+	info      bool
+	data      bool
+	close     bool
+}
+
+func (mc *mockServer) CreateSystemMetricRecord(_ context.Context, in *internal.SystemMetrics) (*internal.SystemMetricsResponse, error) {
+	if mc.createErr {
+		return nil, errors.New("create error")
+	}
+	mc.info = true
+	return &internal.SystemMetricsResponse{
+		Id: "ID",
+	}, nil
+}
+
+func (mc *mockServer) AddSystemMetrics(_ context.Context, in *internal.SystemMetricsData) (*internal.SystemMetricsResponse, error) {
+	if mc.addErr {
+		return nil, errors.New("add error")
+	}
+	mc.data = true
+	return &internal.SystemMetricsResponse{
+		Id: "ID",
+	}, nil
+}
+
+func (mc *mockServer) StreamSystemMetrics(internal.CedarSystemMetrics_StreamSystemMetricsServer) error {
+	return nil
+}
+
+func (mc *mockServer) CloseMetrics(_ context.Context, in *internal.SystemMetricsSeriesEnd) (*internal.SystemMetricsResponse, error) {
+	if mc.closeErr {
+		return nil, errors.New("close error")
+	}
+	mc.close = true
+	return &internal.SystemMetricsResponse{
+		Id: "ID",
+	}, nil
+}
+
+func TestNewSystemMetricsClient(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	srv := &mockServer{}
+	require.NoError(t, startRPCService(ctx, srv, 5000))
+	t.Run("ValidOptions", func(t *testing.T) {
+		connOpts := ConnectionOptions{
+			Client: http.Client{},
+			DialOpts: timber.DialCedarOptions{
+				BaseAddress: "localhost",
+				RPCPort:     "5000",
+			},
+		}
+		client, err := NewSystemMetricsClient(ctx, connOpts)
+		require.NoError(t, err)
+		require.NotNil(t, client)
+		require.NoError(t, client.CloseSystemMetrics(ctx, "ID"))
+		assert.True(t, srv.close)
+	})
+	t.Run("InvalidOptions", func(t *testing.T) {
+		connOpts := ConnectionOptions{}
+		client, err := NewSystemMetricsClient(ctx, connOpts)
+		require.Error(t, err)
+		require.Nil(t, client)
+	})
+}
+
+func TestNewSystemMetricsClientWithExistingClient(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	srv := &mockServer{}
+	require.NoError(t, startRPCService(ctx, srv, 6000))
+	addr := fmt.Sprintf("localhost:%d", 6000)
+	conn, err := grpc.DialContext(ctx, addr, grpc.WithInsecure())
+	require.NoError(t, err)
+
+	t.Run("ValidOptions", func(t *testing.T) {
+		client, err := NewSystemMetricsClientWithExistingConnection(ctx, conn)
+		require.NoError(t, err)
+		require.NotNil(t, client)
+		require.NoError(t, client.CloseSystemMetrics(ctx, "ID"))
+		assert.True(t, srv.close)
+	})
+	t.Run("InvalidOptions", func(t *testing.T) {
+		client, err := NewSystemMetricsClientWithExistingConnection(ctx, nil)
+		require.Error(t, err)
+		require.Nil(t, client)
+	})
+}
+
+func TestCloseClient(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	srv := &mockServer{}
+	require.NoError(t, startRPCService(ctx, srv, 7000))
+	t.Run("WithClientConn", func(t *testing.T) {
```

Comment:
I guess the name `clientConn` is right according to implementation details, but would it be better to for this test and the one below it be swapped/renamed since just reading the test name implies that it will provide a client as input?